shared/mad/codeql/mad/dynamic/GraphExport.qll

```diff
@@ -0,0 +1,230 @@
+   * to be named with a type name.
+   *
+   * Should not hold for nodes that are named via `exposedName`.
+   */
+  predicate suggestedName(Node node, string name);
+
+  /**
+   * Holds if `node` must be named if it part of the exported graph.
+   */
+  predicate mustBeNamed(Node node);
+
+  /**
+   * Holds if the exported graph should contain `node`, if it is reachable from an exposed node.
+   *
+   * This ensures that all paths leading from an exposed node to `node` will be exported.
+   */
+  predicate shouldContain(Node node);
+
+  /**
+   * Holds if `host` has a method that returns `this`, and `path` is the path from `host`
+   * to the method followed by the appropriate `ReturnValue` token.
+   *
+   * For example, if the method is named `m` then `path` should be `Member[m].ReturnValue`
+   * or `Method[m].ReturnValue`.
+   */
+  bindingset[host]
+  predicate hasTypeSummary(Node host, string path);
+
+  /**
+   * Holds if paths going through `node` should be blocked.
+   *
+   * For example, this can be the case for functions that are public at runtime
+   * but intended to be private.
+   */
+  predicate shouldNotContain(Node node);
+}
+
+/**
+ * Module for exporting an arbitrary graph as models-as-data rows.
+ */
+module GraphExport<NodeSig Node, GraphExportSig<Node> S> {
+  private import S
+
+  private Node getAnExposedNode() {
+    not shouldNotContain(result) and
+    (
+      exposedName(result, _, _)
+      or
+      edge(getAnExposedNode(), _, result)
+    )
+  }
+
+  pragma[nomagic]
+  private predicate exposedEdge(Node pred, string path, Node succ) {
+    // Materialize this relation so we can access 'edge' without binding set on 'pred'
+    pred = getAnExposedNode() and
+    edge(pred, path, succ)
+  }
+
+  private Node getARelevantNode() {
+    result = getAnExposedNode() and
+    shouldContain(result)
+    or
+    exposedEdge(result, _, getARelevantNode())
+  }
+
+  final private class FinalNode = Node;
+
+  private class RelevantNode extends FinalNode {
+    RelevantNode() { this = getARelevantNode() }
+  }
+
+  pragma[inline]
+  private RelevantNode getAPredecessor(RelevantNode node, string path) {
+    exposedEdge(result, path, node)
+  }
+
+  private predicate nodeMustBeNamed(RelevantNode node) {
+    exposedName(node, _, "")
+    or
+    S::mustBeNamed(node)
+    or
+    strictcount(getAPredecessor(node, _)) > 1
+  }
+
+  /** Gets a type-name to use as a prefix, in case we need to synthesize a name. */
+  private string getAPrefixTypeName(RelevantNode node) {
+    result =
+      min(string prefix |
+        exists(string type, string path |
+          exposedName(node, type, path) and
+          prefix = join(type, path)
+        )
+        or
+        suggestedName(node, prefix)
+      |
+        prefix
+      )
+    or
+    not exposedName(node, _, _) and
+    not suggestedName(node, _) and
+    result = getAPrefixTypeName(getAPredecessor(node, _))
```

Comment:
This is needed for synthesising type names that give some indicator of where it came from. So instead of a type called `expr123` we get something like `mysql.Connection~expr123`. Synthesised names are still somewhat hard to read but this is better than nothing.

Reply:
Another alternative is that we could also pick one of the access paths leading to the synthesised name to use instead of `expr123`. That would also make it less dependent on locations. So instead of `mysql.Connection~Member[on].Argument[1]`.

Reply:
Making this as 'resolved' based on the "👍" reaction above